### PR TITLE
Issue #1141 enables thin cluster cache configuration

### LINKED ISSUE #1141: Enable "thin" cluster cache configuration
This issue is a companion to and depends on issue #1138.

Defining a clustered cache presently requires either the presence of a `/ehcache:config/ehcache:cache/ehcache:resources/tc:cluster-shared` or `/ehcache:config/ehcache:cache/ehcache:resources/tc:cluster-fixed` element in an XML configuration or the `ResourcePoolsBuilder.with(ClusteredResourcePoolBuilder.<type>)` where `<type>` is either `fixed` or `shared` for programmatic configuration.  And, depending on the configuration values, the clustered service configuration must be available.

For a "thin" clustered cache configuration, the cache must still be marked as _clustered_ but the fixed or shared resource should be omittable.  What is proposed here is the addition of a `clustered` resource type in addition to the `cluster-fixed` and `cluster-shared` resource types.

Support for this requires addition of a `/ehcache:config/ehcache:cache/ehcache:resources/tc:clustered` element to the XSD along with supporting changes in `ClusteredResourceConfigurationParser` and matching changes `ClusteredResourcePoolBuilder`.  An extension of `ClusteredResourcePool` is needed along with a new `ClusteredResourceType`.  A new `ServerStoreConfiguration.PoolAllocation` _may_ be desired.  These additions cascade into other clustering classes.

The validation performed by `EhcacheActiveEntity.validateServerStore` needs to be changed to accept the new _clustered_ pool type.

Although not currently needed to support a "thin" clustered cache configuration, the `EhcacheClientEntity.validateCache` method (and underlying mechanism) should be changed to support return of the `ServerStoreConfiguration` in use by the server (to be consistent with the change to `EhcacheClientEntity.validate` in issue #1138).

### PR DESCRIPTION
**Pull request by gibsong**
Fixes #1141 removes potential null ptr exception and updates javadoc

### CODE REVIEW

**Comment by gibsong** [comment]
Note: I did not implement EhcacheClientEntity.validateCache to return ServerStoreConfiguration

File: clustered/client/src/main/java/org/ehcache/clustered/client/config/ThinClusteredResourcePool.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ehcache.clustered.client.config;
+
+/**
+ *
+ * @author GGIB
```
Comment (ljacomet) [inline clustered/client/src/main/java/org/ehcache/clustered/client/config/ThinClusteredResourcePool.java]: Please drop `@author` tag, we are not using them

File: clustered/client/src/main/java/org/ehcache/clustered/client/config/builders/ClusteredResourcePoolBuilder.java [hunk 2]
```diff
@@ -62,4 +64,11 @@ public static FixedClusteredResourcePool fixed(String fromResource, long size, M
   public static SharedClusteredResourcePool shared(String sharedResource) {
     return new SharedClusteredResourcePoolImpl(sharedResource);
   }
+
+  /**
+   * Creates a resource pool that inherits the resources already configured on the server.
+   */
+  public static ThinClusteredResourcePool thin() {
```
Comment (ljacomet) [inline clustered/client/src/main/java/org/ehcache/clustered/client/config/builders/ClusteredResourcePoolBuilder.java]: I really dislike that name. What about `clustered()` simply?

File: clustered/client/src/main/java/org/ehcache/clustered/client/internal/config/xml/ClusteredResourceConfigurationParser.java [hunk 3]
```diff
@@ -84,6 +85,10 @@ public ResourcePool parseResourceConfiguration(final Element fragment) {
 
       return new FixedClusteredResourcePoolImpl(from, size, sizeUnits);
     }
+    else if("clustered".equals(elementName))
+    {
```
Comment (ljacomet) [inline clustered/client/src/main/java/org/ehcache/clustered/client/internal/config/xml/ClusteredResourceConfigurationParser.java]: Code style

File: clustered/client/src/test/java/org/ehcache/clustered/client/internal/config/xml/ClusteringServiceConfigurationParserTest.java [hunk 4]
```diff
@@ -27,7 +27,6 @@
 import javax.xml.parsers.DocumentBuilder;
 import javax.xml.parsers.DocumentBuilderFactory;
 import javax.xml.transform.stream.StreamSource;
-
```
Comment (ljacomet) [inline clustered/client/src/test/java/org/ehcache/clustered/client/internal/config/xml/ClusteringServiceConfigurationParserTest.java]: Get rid of that totally unrelated change

File: clustered/client/src/test/resources/configs/thin-cluster-cache-invalid-element.xml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  <eh:service>
+    <tc:cluster>
+      <tc:connection url="terracotta://example.com/myCacheManager`+"/>
+      <tc:server-side-config auto-create="false"/>
+    </tc:cluster>
+  </eh:service>
+
+  <eh:cache alias="thinCache">
+    <eh:resources>
+      <tc:clustered>xxx</tc:clustered>
```
Comment (ljacomet) [inline clustered/client/src/test/resources/configs/thin-cluster-cache-invalid-element.xml]: clearly, if we use `clustered` in XML and `thin` in code, there is a name issue.

**Comment by chrisdennis** [comment]
A number of fractured side discussions between @GaryWKeim @cljohnso @AbfrmBlr @ljacomet and myself led to the idea of "clustered", "clustered-shared" and "clustered-dedicated" as suitable terminology for the three clustered resource types.  If nobody has any objections I suggest that we standardize on those names as part of this PR.

File: clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java [hunk 6]
```diff
@@ -76,6 +77,26 @@ public void verify(ServerStoreConfiguration serverConfiguration, ServerStoreConf
     if (!isCompatible) {
       throw new ClusteredStoreValidationException(sb.toString());
     }
+
+  }
+
+  private static boolean comparePoolAllocationType(StringBuilder sb, PoolAllocation serverPoolAllocation, PoolAllocation clientPoolAllocation) {
+    if ((serverPoolAllocation == null && clientPoolAllocation == null)
```
Comment (chrisdennis) [inline clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java]: What does it mean for both to be null?  Surely this method should never be called with both being null?

File: clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java [hunk 7]
```diff
@@ -42,27 +42,28 @@ public void verify(ServerStoreConfiguration serverConfiguration, ServerStoreConf
 
     PoolAllocation serverPoolAllocation = serverConfiguration.getPoolAllocation();
     PoolAllocation clientPoolAllocation = clientConfiguration.getPoolAllocation();
```
Comment (chrisdennis) [inline clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java]: It feels to me like the methods of ServerStoreCompatibility should really be instance methods of the various types it validates.  Does that make sense?

File: clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreConfiguration.java [hunk 8]
```diff
@@ -159,6 +159,13 @@ public String getResourcePoolName() {
         return resourcePoolName;
       }
     }
+
+    /**
+     * Creates a Pool Allocation which inherits the Shared or Fixed Pool Allocation from a cache which is already configured on the server.
+     */
+    final class Thin implements PoolAllocation {
```
Comment (chrisdennis) [inline clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreConfiguration.java]: This class (and the other inner classes here) should probably all be static, right?

File: clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java (hunk 7 above)
Comment (gibsong) [inline clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java]: yes that makes sense.   I will try to refactor code into instance methods.  This should help clean it up a bit too.

File: clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java (hunk 6 above)
Comment (gibsong) [inline clustered/common/src/main/java/org/ehcache/clustered/common/ServerStoreCompatibility.java]: I removed the null checks.  It appears that this case can't happen.  If I am wrong please correct me.

File: clustered/client/src/main/java/org/ehcache/clustered/client/config/builders/ClusteredResourcePoolBuilder.java [hunk 9]
```diff
@@ -32,25 +34,25 @@ private ClusteredResourcePoolBuilder() {
...
-   * @param fromResource the name of the server-based resource from which this fixed resource pool
-   *                     is reserved; may be {@code null}
+   * @param fromResource the name of the server-based resource from which this dedicated resource pool
+                     is reserved; may be {@code null}
    * @param size       the size
    * @param unit       the unit for the size
    */
-  public static FixedClusteredResourcePool fixed(String fromResource, long size, MemoryUnit unit) {
-    return new FixedClusteredResourcePoolImpl(fromResource, size, unit);
+  public static DedicatedClusteredResourcePool dedicated(String fromResource, long size, MemoryUnit unit) {
```
Comment (ljacomet) [inline clustered/client/src/main/java/org/ehcache/clustered/client/config/builders/ClusteredResourcePoolBuilder.java]: I am in favour of `clusteredDedicated` here

File: clustered/client/src/main/java/org/ehcache/clustered/client/config/builders/ClusteredResourcePoolBuilder.java [hunk 10]
```diff
@@ -62,4 +64,11 @@ public static FixedClusteredResourcePool fixed(String fromResource, long size, M
   public static SharedClusteredResourcePool shared(String sharedResource) {
```
Comment (ljacomet) [inline clustered/client/src/main/java/org/ehcache/clustered/client/config/builders/ClusteredResourcePoolBuilder.java]: I am in favour of `clusteredShared` here

**Comment by ljacomet** [comment]
test this please

**Comment by ljacomet** [comment]
Verified this runs tests and passes checks locally so merging.
